doc/phar-signing.md

```diff
@@ -0,0 +1,364 @@
+
+There is two idiomatic ways to secure a PHAR:
+
+- Using the built-in PHAR signing API (**not** recommended; read [Why it is bad](#why-it-is-bad) for the why).
+- Signing the PHAR as any other generic binary (see [Sign your PHAR](#sign-your-phar)).
+
+This doc entry goal is to show why the first method is to be avoided and how to do it "the right way".
+
+
+## Built-in PHAR API
+
+### How to sign your PHAR
+
+This is how a PHAR can be signed:
+
+```php
+// See https://www.php.net/manual/en/phar.setsignaturealgorithm.php
+$phar->setSignatureAlgorithm($algo, $privateKey);
+```
+
+There is various algorithm available. The most "secure" one would be `Phar::OPENSSL` with an
+OpenSSL private key. For instance:
+
+```
+$ openssl genrsa -des3 -out acme-phar-private.pem 4096
+```
+
+```php
+// E.g. $privateKeyPath = 'acme-phar-private.pem' with the example above
+$privateKey = file_get_contents($privateKeyPath);
+
+$resource = openssl_pkey_get_private($key, $privateKeyPassword);
+openssl_pkey_export($resource, $private);
+$details = openssl_pkey_get_details($resource);
+
+$phar->setSignatureAlgorithm(Phar::OPENSSL, $private);
+
+file_put_contents(
+    $phar->getPath().'.pubkey',
+    $details['key'],
+);
+```
+
+With the example above, you will end up with two files: your PHAR (e.g. `bin/command.phar`) and its
+public key (e.g. `bin/command.phar.pubkey`).
+
+
+### How it works
+
+To give more background on how PHAR archives are constructed, they are PHP files that contain a mixture of code and
+other binary data. It is analogous to the [JAR files][jar], and will require PHP to execute the PHARs. The content
+of a PHAR is as follows:
+
+1. A stub: a piece of code that handles the extraction of resources.
+1. A binary manifest which allows the interpreter to understand the file structure of the embedded contents that follow.
+1. The actual content of the archive.
+1. The signature makes up the last section of the file. It is a 4-byte signature flag which tells what signature type
+   was used and then another 4-byte constant marks the file as having a signature.
+
+When PHP later reads the archive, it can determine the signature and type by reading the end of the archive. This way,
+if the content of the PHAR has been tempered with, e.g. code was injected in the archive, PHP will see that the content
+of the archive does not match the signature of the PHAR and will bail out.
+
+
+### Why it is bad
+
+If a PHAR gets corrupted, the PHP interpreter will know from the signature and throw a `PharException`.
+
+The first issue is the signature itself: if the PHAR gets corrupted, maybe the signature got corrupted too. Using an
+OpenSSL based signature will help to prevent this issue however.
+
+The second issue is the mechanism itself. Indeed, if one injects code _before_ the stub, then this code will be executed
+before the signature check.
+
+**This security mechanism cannot prevent modifications of the archive itself.**  
+
```

Comment:
There are more bad aspects of the built-in signing solution:

- You cannot run the phar without the pub key laying right next to it - which is just stupid and violates the unix filesystem standards when you place the `command.phar` into something like `/usr/local/bin`.
- OpenSSL keys do not contain any identity information. So unless cleanly separated at distribution time, nobody knows where the pub key came from or who generated it. Which (almost) kills the very idea of signing things.

With this design choice, the full process of verifying the signature falls short. You can just make a new signature and place a new pub key next to the phar and nobody will have any means of noticing. That's why for GPG signatures, you have to import the pub key into your keyring first and it's linked to an identity - and your explicit choice to trust said entity.